### ISSUE TITLE
(inactive) CustomGenome

### ISSUE THREAD

**Issue by mtekman** [Bioconductor/Contributions#3224]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/mtekman/CustomGenome

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by jorainer** [comment]
Hi, I'm just providing some thoughts/comments since I was also dealing a lot with Ensembl annotations.

1) you are downloading fasta and GTF sequences yourself from Ensembl - have you checked whether these would also be available in Bioconductor's `AnnotationHub`? Or, would it maybe make sense to use the `BiocFileCache` package for downloading/caching these resources? IMHO that would simplify this process.

The general idea to create *custom* annotations or add additional sequence/gene information to reference genomes is however really nice.

**Comment by mtekman** [comment]
Hello @jorainer, thanks for the comments!

> The general idea to create custom annotations or add additional sequence/gene information to reference genomes is however really nice.

Exactly, I was surprised it wasn't already a package!

*******

> AnnotationHub

Ah that's why these reviews are golden. I wish I had heard about this before I implemented the backend for Ensembl downloads that I did.

```R
library(AnnotationHub)
ah = AnnotationHub(cache = "/my/dest/cache")
ah[ah$species == "Mus musculus" &
     ah$dataprovider == "Ensembl" &
     ah$genome=="GRCm38" &
     ah$sourcetype %in% c("GTF", "ensembl")]
```

The problem I'm seeing is how to pair a given genome with a given GTF file? With Ensembl this is pretty unambiguous, but with AnnotationHub I'm overwhelmed with choice.

I think this is something I will implement into the next version of this tool, since right now this is very much geared towards Ensembl downloads.

*******

> BiocFileCache

I had never heard of this, and it sounds pretty useful for some future analyses. I have some angst related to how it checks for data consistency:

https://bioconductor.org/packages/release/bioc/vignettes/BiocFileCache/inst/doc/BiocFileCache.html

> > bfcneedsupdate() is a method that will check the local copy of the data’s etag and last_modifed time to the etag and last_modified time of the remote resource as well as an expires time. The cache saves this information when the web resource is initially added. The expires time is checked against the current Sys.time to see if the local resource has expired. If so the resource will deem need to be updated; if unavailable or not expired will check the etag and last_modified_time. The etag information is used definitively if it is available, if it is not available it checks the last_modified time. If the resource does not have a last_modified tag either, it is undetermined. If the resource has not been download yet, it is TRUE

It sounds like it doesn't do any consistency checks with remote checksums, which is a main indicator for me on whether a large file downloaded isn't truncated. Have you had any issues with this in your code?

*********

Also, how does BiocFileCache interact with AnnotionHub? I have the impression that AnnotationHub has it's own caching.

**Comment by lshep** [comment]
BiocFileCache is used as the backend caching mechanism in AnnotationHub.

**Comment by lshep** [comment]
You'll need an `inst/script` data that describes how the data in `inst/extdata` was generated, and include any relevant source and licensing information for the data there.

**Comment by lshep** [comment]
Please correct ERROR, Warning, Notes from the build report before a reviewer is assigned for in depth review.

**Comment by mtekman** [comment]
@lshep  Thanks, I've fixed the issue! Sorry for the long delay, it was a busy month

**Comment by lshep** [comment]
Please push the changes to git.bioconductor.org to trigger a new build report.  see https://github.com/Bioconductor/Contributions/issues/3224#issuecomment-1819369522

**Comment by mtekman** [comment]
@lshep I'm not sure how to get around these errors. The main problem with them is that sometimes a file downloads incomplete and then the checksum fails, but sometimes this works fine (as is the case here: https://github.com/Bioconductor/Contributions/issues/3224#issuecomment-1870657807)

What do I need to do to get this to pass, other than removing more test checks that work fine on my local devices?

**Comment by lshep** [comment]
Do you implement any sort of caching so that if a file works it doesnt have to be redownloaded the next time?  See [BiocFileCache](https://bioconductor.org/packages/release/bioc/html/BiocFileCache.html)

**Comment by mtekman** [comment]
I don't, but the code does check to see if the file has already been downloaded to the temp folder. I think the main issue is the sheer size of these files.

Using BiocFileCache wouldn't fix this problem if I understand correctly, since if I run this N times over N push/commits

```R
path <- tempfile()
bfc <- BiocFileCache(path, ask = FALSE)
bfcadd(bfc, "genome", fpath="https://www.ensembl.com/big/file.fasta.gz")
bfcadd(bfc, "othergenome", fpath="https://www.ensembl.com/big/file2.fasta.gz")
## later:
mygenome1 <- bfcquery(bfc, "genome")
mygenome2 <- bfcquery(bfc, "othergenome")
```
it would *still* initialise N different independent caches and still have to initially download all files that the start of each test.

This doesn't change much from my initial code which downloads to a fixed folder for genomes and re-uses the resources over subsequent calls.

Is there anything else I can do?

**Comment by lshep** [comment]
If the size of the files are large then caching will be a necessity and likely a requirement by a reviewer.  I would recommend a package specific directory that we general allow in cases like this.  See UniProt.ws example:  
```
    cache <- tools::R_user_dir("UniProt.ws", "cache")
    bfc <- BiocFileCache(cache, ask=FALSE)
```
But also be very clear to the user how much space they will need to run example data / vignettes before the start -- understanding that it can be variable once used with other data.

**Comment by mtekman** [comment]
This version implements BiocFileCache (within get_genome_files) and retires
the manual checksum based methods of confirming a cached file.

I see some warnings, but otherwise I think it's ready to be reviewed again.

**Comment by PeteHaitch** [comment]
Hi @mtekman,

I started to review **CustomGenome** but it soon became clear that it needs some improvements before the review can proceed.

I always start my package review by working through the vignette, like a new user would when they encounter your package.
The current vignette has almost no evaluated code chunks, as noted by `BiocCheck()`:

```r
BiocCheck::BiocCheck('CustomGenome_0.99.10.tar.gz')
# Omitting some output
* Checking vignette directory...
    * WARNING: Evaluate more vignette chunks.
        10 out of 13 code chunks = 76% unevaluated
        10 non-exec code chunk(s) (e.g., '```r')
# Omitting some more output
```

Instead, the vignette has static copy+pasted code and output, and the problem with that is it can easily become out of sync with the actual code in the package.
Even the evaluated code chunks, like the very first one, highlight that the documentation has fallen out-of-sync with the package code:

```r
library(CustomGenome)
data("user_sequences")
#> Warning in data("user_sequences"): data set 'user_sequences' not found
```

Continuing on, the next chunk contains invalid code:

```r
homo_sap = get_genome_files(species="homo_sapiens",
                            output_folder="genomes")
#> Error in get_genome_urls(...): unused argument (output_folder = "genomes")
```

I corrected that error (using `cache_folder` instead of `output_folder`) and was then surprised to find that the code started downloading the genome files for Mus musculus rather than Homo sapiens!

```r
homo_sap = get_genome_files(species="homo_sapiens",
                            cache_folder="genomes")
#> Downloading: Mus_musculus.GRCm39.105.gtf.gz                              
```

At this point I stopped my review because it's clear the package is not yet ready.

A few other comments from what I saw when reading the first part of the vignette:

- The vignette must mostly comprise evaluated code chunks. I appreciate that this package needs to download and process relatively large files, so you may need to switch the examples in the vignette and documentation to smaller (or subsetted) genomes to make this feasible within Bioconductor's time and resource requirements for building and checking packages.
- The vignette should use [**BiocStyle**](https://bioconductor.org/packages/release/bioc/html/BiocStyle.html) for formatting.
- We recommend using `<-` rather than `=` for variable assignment.
- As [Lori explained](https://github.com/Bioconductor/Contributions/issues/3224#issuecomment-1883026597), the default `cache_folder` value of`get_genome_files()` should be `tools::R_user_dir("CustomGenome", which = "cache")` rather than `genomes`
- The formatting in the documentation (`.Rd` files in the `man/` directory) is a bit messy (e.g., see `?get_genome_files`), which I think because the markdown formatting is incorrect in the corresponding roxygen lines in the `.R` files.

Please take a read of https://contributions.bioconductor.org/ to ensure that your package complies with this advice.
I'm going to mark this as 'pending pre-review changes'.
We can restart the review once you have addressed these issues and confirmed that you've read https://contributions.bioconductor.org/ to ensure that your package complies with the advice therein.

Cheers,
Pete

**Comment by mtekman** [comment]
Hi @PeteHaitch 

Thanks a lot for the review, I guess I need to work a lot on the vignette. I somehow assumed that a reproducible vignette was not a neccessity, especially for large datasets -- but I can see that I neglected it so long that it did indeed fall out of sync with the code.

Cheers to @lshep  for the R_user_dir suggestion, I somehow overlooked that comment when I reimplemented the caching.

I'll get back on this soon, thanks again
